File: packages/protocol/src/events.ts
/**
 * Unified agent events — both Codex SDK events and Claude stream-json events
 * are mapped to these types before being sent to the mobile client.
 */

import type { AgentState, FileChange, TokenUsage } from "./state.js";

// ─── Agent → Phone events ────────────────────────────────────────────

export interface StatusEvent {
  type: "status";
  state: AgentState;
  message: string;
}

export interface ThinkingEvent {
  type: "thinking";
  text: string;
}

export interface CodeChangeEvent {
  type: "code_change";
  changes: FileChange[];
}

export interface CommandExecEvent {
  type: "command_exec";
  command: string;
  output?: string;
  exitCode?: number;
  status: "running" | "done" | "failed";
}

export interface AgentMessageEvent {
  type: "agent_message";
  text: string;
}

export interface ErrorEvent {
  type: "error";
  message: string;
}

export interface TurnCompletedEvent {
  type: "turn_completed";
  summary: string;
  filesChanged: string[];
  usage: TokenUsage | null;
}

/**
 * Union of all possible agent events.
 */
export type AgentEvent =
  | StatusEvent
  | ThinkingEvent
  | CodeChangeEvent
  | CommandExecEvent
  | AgentMessageEvent
  | ErrorEvent
  | TurnCompletedEvent;

File: crates/codepilot-protocol/src/events.rs
use serde::{Deserialize, Serialize};

use crate::state::{AgentState, FileChange, TokenUsage};

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "lowercase")]
pub enum CommandExecStatus {
    Running,
    Done,
    Failed,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum AgentEvent {
    Status {
        state: AgentState,
        message: String,
    },
    Thinking {
        text: String,
    },
    CodeChange {
        changes: Vec<FileChange>,
    },
    CommandExec {
        command: String,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        output: Option<String>,
        #[serde(default, rename = "exitCode", skip_serializing_if = "Option::is_none")]
        exit_code: Option<i64>,
        status: CommandExecStatus,
    },
    AgentMessage {
        text: String,
    },
    Error {
        message: String,
    },
    TurnCompleted {
        summary: String,
        #[serde(rename = "filesChanged")]
        files_changed: Vec<String>,
        #[serde(default)]
        usage: Option<TokenUsage>,
    },
}
